### ISSUE TITLE
Perfomance drop on ONNX compared to PyTorch

### ISSUE THREAD

**Issue by kkeroo** [ultralytics/ultralytics#20930]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Hi, I trained the YOLOv8n model using ultralytics on Global Wheat Head Dataset 2021 and got 0.5 mAP@50-95. I didnt change the `confidence` or `iou` or something that could affect the nms. Here is the full command:
```
results = model.train(data=path_to_dataset_config, epochs=200, imgsz=(768,768), batch=16, degrees=20, flipud=0.1, patience=40, mosaic=0.5, name="yolov8n-768", single_cls=True, box=10.0, cls=0.1, dfl=0.1)
``` 

I once again did the validation using the `best.pt` like
```
metrics = model.val(data=path_to_dataset_config, imgsz=(768,768), batch=16)
```
and got the same mAP results. Then I exported the model to onnx file (tried different opset version and always got the same results):
```
model = YOLO("runs/detect/yolov8n-768/weights/best.pt")
model.export(format="onnx", imgsz=(768,768))
```

And then I did the validation again using this onnx file
```
model = YOLO("runs/detect/yolov8n-768/weights/best.onnx")
metrics = model.val(data=path_to_dataset_config, imgsz=(768,768), batch=16)
```

And the results are not the same although they should be (getting only 0.46 mAP@50-95). Or at least be more similar. I am aware that a slight performance drop could be in ONNX using FP16 but I also tried setting `half=False`. 

Is there any other reason why this happens? Should I try to set different `conf` and `iou` values when validating? I did that and got some better results setting the `conf=0.3`. 

What are the default values for confidence scores and ious that are used? Could this be the reason for different performance?

Thanks a lot!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @kkeroo, thank you for bringing this up and for your detailed description! 🚀 This is an automated response to help get you started while an Ultralytics engineer reviews your question.

We recommend visiting the [Docs](https://docs.ultralytics.com/) for guidance on [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) examples, as well as answers to many common questions.

Since this may be a 🐛 Bug Report, could you please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) (MRE)? This will help us investigate and debug the issue more efficiently.

If you are experimenting with custom training or export settings, please include as much detail as possible, such as dataset image samples, training logs, and validation outputs. Also, check out our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community for real-time chat on [Discord](https://discord.com/invite/ultralytics) 🎧, in-depth discussions on [Discourse](https://community.ultralytics.com/), or to share insights on our [Subreddit](https://reddit.com/r/Ultralytics).

## Upgrade

Please ensure you are using the latest `ultralytics` package and all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

YOLO runs in these up-to-date, verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

An Ultralytics engineer will review your issue soon. Thank you for your patience and for helping improve Ultralytics! 🦾

**Comment by Y-T-G** [comment]
You didn't export with `batch=16`, so it runs with `batch=1`.

**Comment by kkeroo** [comment]
Hm this shouldn't affect the mAP?

**Comment by Y-T-G** [comment]
There's also a warning that says:

```
WARNING ⚠️ updating to 'imgsz=640'. 'train' and 'val' imgsz must be an integer, while 'predict' and 'export' imgsz may be a [h, w] list or an integer, i.e. 'yolo export imgsz=640,480' or 'yolo export imgsz=640'
```

which means the PyTorch model doesn't run validation at square image but the ONNX model does (because the imgsz is fixed for ONNX model since you didn't export with `dynamic=True`).

You should get the same mAP if you run both validation with `rect=False` or export the ONNX model with `dynamic=True`.

**Comment by kkeroo** [comment]
I exported the model with
```
model.export(format="onnx", imgsz=(768,768), dynamic=True)
```
and validated the onnx with:
```
metrics = model.val(data=dataset_path, half=False, imgsz=[768,768], rect=False)
```
and I am getting the same mAP drop.

**Comment by kkeroo** [comment]
note that images from GWHD dataset are rectangular (1024x1024) so no padding is done probably

**Comment by kkeroo** [comment]
Ok, and when I validate the .pt model using `rect=False` I get the same mAP drop as with ONNX. So my question is can I achieve the same mAP with ONNX as with .pt and `rect=True`?

**Comment by Y-T-G** [comment]
Yes if you exported with dynamic=True, then run the validation with the same arguments as PyTorch (rect=True and batch=16)

**Comment by kkeroo** [comment]
Why would batch size affect the results? If I validate pytorch model with batch size 1 I get the same results as with batch size 16? Nevertheless I tried your approach and it does not solve the issue. Any other ideas?

**Comment by Y-T-G** [comment]
Batch size affects result with `rect=True` and if your images have different aspect ratios. It doesn't affect with `rect=False` and if all image has same aspect ratio. It's because of how `rect` works.

**Comment by Y-T-G** [comment]
> Nevertheless I tried your approach and it does not solve the issue. Any other ideas?

What command did you run?

**Comment by Y-T-G** [comment]
It's the same for me if I export with `dynamic=True`:

```python
In [2]: results = model.val(data="coco128.yaml", exist_ok=True, batch=16, device=-1)
Searching for 1 idle GPUs with free memory >= 20.0% and free utilization >= 0.0%...
Selected idle CUDA devices [0]
Ultralytics 8.3.148 🚀 Python-3.11.10 torch-2.7.0+cu126 CUDA:0 (NVIDIA A100-SXM4-80GB, 81038MiB)
YOLO11n summary (fused): 100 layers, 2,616,248 parameters, 0 gradients, 6.5 GFLOPs
val: Fast image access ✅ (ping: 0.0±0.0 ms, read: 1425.7±546.2 MB/s, size: 45.8 KB)
val: Scanning /datasets/coco128/labels/train2017.cache... 126 images, 2 backgrounds, 0 corrupt: 100%|██████████| 128/128
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 8/8 [00:02<00
                   all        128        929      0.662       0.59      0.671      0.505

In [4]: model = YOLO("yolo11n.onnx")
WARNING ⚠️ Unable to automatically guess model task, assuming 'task=detect'. Explicitly define task for your model, i.e. 'task=detect', 'segment', 'classify','pose' or 'obb'.

In [5]: results = model.val(data="coco128.yaml", exist_ok=True, batch=16, device=-1)
Searching for 1 idle GPUs with free memory >= 20.0% and free utilization >= 0.0%...
Selected idle CUDA devices [1]
Ultralytics 8.3.148 🚀 Python-3.11.10 torch-2.7.0+cu126 CUDA:1 (NVIDIA A100-SXM4-80GB, 81038MiB)
Loading yolo11n.onnx for ONNX Runtime inference...
Using ONNX Runtime CUDAExecutionProvider
val: Fast image access ✅ (ping: 0.0±0.0 ms, read: 1336.5±592.0 MB/s, size: 54.4 KB)
val: Scanning /datasets/coco128/labels/train2017.cache... 126 images, 2 backgrounds, 0 corrupt: 100%|██████████| 128/128
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 8/8 [00:05<00
                   all        128        929      0.662       0.59      0.671      0.505
```

**Comment by kkeroo** [comment]
I tried all options now and made a MRE. I used your yolov8 and `GlobalWheat2020.yaml` dataset.

Trained the model for one epoch with:
```
from ultralytics import YOLO
import os
import torch
model = YOLO("yolov8n.yaml")  # load a pretrained model (recommended for training)
path_to_dataset_config = "GlobalWheat2020.yaml"
results = model.train(data=path_to_dataset_config, epochs=125, imgsz=640, batch=8)
```

Then I validated the model like:
```
from ultralytics.utils.benchmarks import benchmark
from ultralytics import YOLO
path_to_dataset_config = "GlobalWheat2020.yaml"
model = YOLO("runs/detect/train/weights/best.pt")
metrics = model.val(data=path_to_dataset_config, imgsz=640)
```

And got 0.674 mAP@50 and 0.28 mAP@50-95. I validated the model using `batch=1` and batch=`16` and got the same results. Then I exported the model with:
```
model = YOLO("runs/detect/train/weights/best.pt")
model.export(format="onnx", imgsz=640, dynamic=True)
```

and benchmarked the model with:
```
path_to_dataset_config = "GlobalWheat2020.yaml"
model = YOLO("runs/detect/train/weights/best.onnx")
metrics = model.val(data=path_to_dataset_config, imgsz=640, rect=True)
```

And got 0.672 mAP@50, 0.278 mAP@50-95.

Could you try the code chunks yourself and verify that your approach is working because I can not get it to work. 

Thanks!

**Comment by Y-T-G** [comment]
I trained for 2 epochs. But the results are the same

```
In [4]: model = YOLO("/ultralytics/runs/detect/train/weights/best.pt")

In [5]: metrics = model.val(data=path_to_dataset_config, imgsz=640, device=-1, bat
   ...: ch=1)
Searching for 1 idle GPUs with free memory >= 20.0% and free utilization >= 0.0%...
Selected idle CUDA devices [1]
Ultralytics 8.3.148 🚀 Python-3.11.10 torch-2.7.0+cu126 CUDA:1 (NVIDIA A100-SXM4-80GB, 81038MiB)
YOLOv8n summary (fused): 72 layers, 3,005,843 parameters, 0 gradients, 8.1 GFLOPs
val: Fast image access ✅ (ping: 0.0±0.0 ms, read: 3429.0±294.0 MB/s, size: 426.2 KB)
val: Scanning /datasets/GlobalWheat2020/labels/ethz_1.cache... 747 images, 0 backg
                 Class     Images  Instances      Box(P          R      mAP50  mAP
                   all        747      49603      0.787      0.701      0.781      0.367
Speed: 0.3ms preprocess, 7.2ms inference, 0.0ms loss, 2.1ms postprocess per image
Results saved to /ultralytics/runs/detect/val

In [8]: model = YOLO("/ultralytics/runs/detect/train/weights/best.onnx")

In [9]: metrics = model.val(data=path_to_dataset_config, imgsz=640, device=-1, batch=1)
Searching for 1 idle GPUs with free memory >= 20.0% and free utilization >= 0.0%...
Selected idle CUDA devices [1]
Ultralytics 8.3.148 🚀 Python-3.11.10 torch-2.7.0+cu126 CUDA:1 (NVIDIA A100-SXM4-80GB, 81038MiB)
Loading /ultralytics/runs/detect/train/weights/best.onnx for ONNX Runtime inference...
Using ONNX Runtime CUDAExecutionProvider
val: Fast image access ✅ (ping: 0.0±0.0 ms, read: 1208.9±941.0 MB/s, size: 371.8 KB)
val: Scanning /datasets/GlobalWheat2020/labels/ethz_1.cache... 747 images, 0 backgrounds, 0 corru
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██
                   all        747      49603      0.787      0.701      0.781      0.367
Speed: 0.3ms preprocess, 7.6ms inference, 0.0ms loss, 3.8ms postprocess per image
Results saved to /ultralytics/runs/detect/val2

```

Maybe you're using an outdated version of Ultralytics. I also get 0.367 mAP50-95 in just 2 epochs starting from `yolov8n.yaml`. But you said you just got 0.278 in 125 epochs.

**Comment by kkeroo** [comment]
No, I trained for only one epoch. I will try updating ultralytics.

**Comment by kkeroo** [comment]
@Y-T-G thanks for help! Upgrading to `ultralytics==8.3.150` (going from `8.3.48`) solved the issue. This is great news.

**Comment by HappyKerry** [comment]
> [@Y-T-G](https://github.com/Y-T-G) thanks for help! Upgrading to `ultralytics==8.3.150` (going from `8.3.48`) solved the issue. This is great news.

need to retrain the model?

**Comment by kkeroo** [comment]
I think no need for retraining, since the discrepancy has to do with postprocessing

**Comment by glenn-jocher** [comment]
No retraining needed—upgrade Ultralytics to the latest, re-export ONNX with dynamic=True, and validate with identical args; defaults are conf=0.001 and iou=0.7. See the Val docs and the ONNX export guide for details: https://docs.ultralytics.com/modes/val/ and https://docs.ultralytics.com/integrations/onnx/.
